pkg/client/introduce/client.go

```diff
@@ -0,0 +1,138 @@
+	"github.com/hyperledger/aries-framework-go/pkg/kms"
+	"github.com/hyperledger/aries-framework-go/pkg/storage"
+)
+
+// provider contains dependencies for the introduce protocol and is typically created by using aries.Context()
+type provider interface {
+	Service(id string) (interface{}, error)
+	KMS() kms.KeyManager
+	InboundTransportEndpoint() string
+	StorageProvider() storage.Provider
+}
+
+// Client enable access to introduce API
+type Client struct {
+	service.Action
+	service.Message
+	service                  service.DIDComm
+	kms                      kms.KeyManager
+	inboundTransportEndpoint string
+	actionCh                 chan service.DIDCommAction
+	msgCh                    chan service.StateMsg
+	store                    storage.Store
+}
+
+// New return new instance of introduce client
+func New(ctx provider) (*Client, error) {
+	svc, err := ctx.Service(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	introduceSvc, ok := svc.(service.DIDComm)
+	if !ok {
+		return nil, errors.New("cast service to Introduce Service failed")
+	}
+
+	store, err := ctx.StorageProvider().OpenStore(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &Client{
+		service:                  introduceSvc,
+		kms:                      ctx.KMS(),
+		inboundTransportEndpoint: ctx.InboundTransportEndpoint(),
+		actionCh:                 make(chan service.DIDCommAction),
+		msgCh:                    make(chan service.StateMsg),
+		store:                    store,
+	}
+
+	// register the action event channel
+	err = c.service.RegisterActionEvent(c.actionCh)
+	if err != nil {
+		return nil, fmt.Errorf("introduce action event registration: %w", err)
+	}
+
+	// register the message event channel
+	err = c.service.RegisterMsgEvent(c.msgCh)
+	if err != nil {
+		return nil, fmt.Errorf("introduce message event registration: %w", err)
+	}
+
+	// start listening for action/message events
+	go c.startServiceEventListener()
+
+	return c, nil
+}
+
+// startServiceEventListener listens to action and message events from introduce service.
+func (c *Client) startServiceEventListener() {
+	// listen for action event and message events
+	for {
+		select {
+		case e := <-c.actionCh:
+			c.ActionEvent() <- e
+		case e := <-c.msgCh:
+			for _, handler := range c.MsgEvents() {
+				handler <- e
+			}
+		}
+	}
+}
+
+func (c *Client) handleOutbound(msg interface{}) error {
+	payload, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	didMsg, err := service.NewDIDCommMsg(payload)
+	if err != nil {
+		return err
+	}
+
+	return c.service.HandleOutbound(didMsg, &service.Destination{})
+}
+
+// CreateProposal creates proposal
+func (c *Client) CreateProposal() error {
+	return c.handleOutbound(&introduce.Proposal{
+		Type: introduce.ProposalMsgType,
+		ID:   uuid.New().String(),
+	})
+}
+
+// CreateSkipProposal creates proposal
+// Skip Proposal should be created when client has public invitation
+func (c *Client) CreateSkipProposal() error {
```

Comment:
This should be removed. There is no "SkipProposal" message type in the RFC. As per our discussion this morning I understood the "skipping" (ie. decrementing wait_count) was going to be handled via other means.

Reply:
Can `introduce.SkipProposalMsgType` be deleted?

Reply:
Done! I am going to delete ` introduce.SkipProposalMsgType` (but not in that PR)